### PR TITLE
Proposal: Lotus / Filecoin RPC and Library Audit

### PR DESCRIPTION
**Pull request by rvagg**
Sudo is currently working on [store-my-text](https://github.com/protocol/web3-dev-team/pull/46) and it's hitting all of the hurty-parts of small-deal making, but it's also raising lots of questions about, and finding problematic issues with, the JavaScript API landscape and the RPC API itself. So this project is intended as a broader RPC API audit as an extension of that work (or replacement of it).

### CODE REVIEW

**Review by vasco-santos — COMMENTED** [review]
Overall this looks good to me.

One of the problems I have seen after playing with [filecoin-shipyard/js-lotus-client](https://github.com/filecoin-shipyard/js-lotus-client) is that the type definitions generated from go do not have into consideration parameters optionality.  This basically means that a typescript project needs to provide all the parameters, even if most (of all of them) are optional, in order to have tsc running. The way the schemas are generated might mean that we do not have a lot of flexibility as the API is basically inferred by generating typescript declarations from lotus api source code.

File: proposals/80-lotus-api-audit.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+   * Other PL staff who may have an interest in, or experience with these libraries and the RPC itself
+ * Attemtping to use the libraries as they currently exist, and possibly the Lotus RPC directly and documenting the process to contribute to the project outputs listed above
+
+#### What does done look like?
+
+_What specific deliverables should completed to consider this project done?_
+
+See list of outputs above.
+
+####  What does success look like?
```
[the remark this answers is not in the document]
Comment (momack2) [reply proposals/80-lotus-api-audit.md:75]: Actually probably a chain of success metrics:
- Ability to rate/sort js lib proposals
- high-value js lib proposals identified and chosen to work on

**Comment by rvagg** [comment]
I've tried to clarify scope here a little better, and fill out some of the additional missing sections.

This project is doing a lot of mapping-of-the-unknown, but it does need to be kept to the minimum required. We currently think we have a decent picture of the landscape so the scope is defined accordingly--although we need to be prepared to discover entirely new universes (I stumbled on the glifio JS stack by accident just prior to writing this proposal!). Where I'm keeping this tight is in the maturity of the output. We're not producing polished docs, we're not even producing docs except where modifications might be trivial activities. We're producing checklists, and have already [started](https://github.com/protocol/w3dt-sudo/issues?q=label%3Afilcoin-rpc-audit) that process. We'll scope further work later but we just need to collect this information by exploring and very basic experimenting.

I've sized it at a Small, but that's for 3-5 FTEs, which is much more than we're working with on Sudo at the moment so I want to set that expectation up front.

@anorth @jacobheun I would appreciate some EM eyes to critique my attempts to define a sense of scope.

**Comment by jacobheun** [comment]
> except where modifications might be trivial activities

I would avoid this regardless of triviality to prevent frequent yet subtle derailments. Reading through this the output seems to best fit a report on the current state of JS clients, and recommendations (stack ranked) for improvements.

Based on the information here it seems that stakeholder interviews should be the first to happen, which would then inform the existing api/client evaluation: Based on what we learned from the stakeholders, what are the libraries we should be focusing our analysis on?

It seems like this is the general path being proposed here, I just want to verify we're not proposing these may be done in parallel. The advantage of doing these in sequence is that we can reduce the footprint of the api we're evaluating and then focus investigation on verifying the observations from the interviews. During validation we may find additional issues that weren't previously raised, but those can be added to the findings of the final report. Overall the scope looks good, but we should ensure to phase it, to avoid splitting focus over too many things.

**Comment by anorth** [comment]
Scope for this project seems appropriately limited. I wouldn't want that to also limit the scope of recommendations that come out of this, though. E.g. maybe the Lotus RPC API that serves node operators and miners isn't the right abstraction, or transport, or contract-to-impose-on-Lotus for client APIs, and work on the libraries that try to solve that impedance mismatch entrenching a long-term problem?

**Comment by rvagg** [comment]
Updated the "done" to say that the goal is to produce a "brief report" covering the state of the world for these libraries and the Lotus RPC API from the library perspective. It should be quite terse and short and aim for maximal information in minimal words and be useful for scoping and prioritising future work and may feed into documentation efforts.

**Review by jacobheun — APPROVED** [review]

**Pull request merged**